perses/analysis/free_energies.py

```diff
@@ -0,0 +1,242 @@
+    details_file_path: str,
+    work_file_path: str,
+    complex_project: str,
+    solvent_project: str,
+    temperature_kelvin: float = 300.0,
+    n_bootstrap: int = 100,
+    show_plots: bool = False,
+    plot_file_format: str = str,
+    cache_dir: Optional[str] = None,
+):
+    r"""Compute free energies from a set of runs.
+
+    Parameters
+    ----------
+    details_file_path : str
+        Path to json file containing run metadata.
+    work_file_path : str
+        Path to bz2-compressed pickle file containing work values from simulation
+    complex_project : str
+        Project identifier (of the form "PROJXXXXX") of the FAH project for complex simulation
+    solvent_project : str
+        Project identifier (of the form "PROJXXXXX") of the FAH project for solvent simulation
+    temperature_kelvin : float
+    n_bootstrap : int, optional
+        Number of bootstrap steps used for BAR free energy estimation
+    show_plots : bool, optional
+        If true, block to display plots interactively (using `plt.show()`).
+        If false, save plots as images in current directory (for batch usage).
+    plot_file_format : str, optional
+        Image file format for saving plots. Accepts any file extension supported by `plt.savefig()`
+    cache_dir : str, optional
+        If given, local directory for caching BAR calculations. Results are cached by run, phase, and generation.
+        If None, no caching is performed.
+    """
+
+    with bz2.BZ2File(work_file_path, "r") as infile:
+        work = pickle.load(infile)
+
+    with open(details_file_path, "r") as f:
+        details = json.load(f)
+
+    work.replace([np.inf, -np.inf], np.nan)
+    work = work.dropna()
+
+    temperature = temperature_kelvin * unit.kelvin
+    kT = kB * temperature
+
+    projects = {"complex": complex_project, "solvent": solvent_project}
+
+    def _produce_plot(name):
+        if show_plots:
+            plt.show()
+        else:
+            plt.savefig('.'.join([name, plot_file_format]))
+
+    def _bootstrap_BAR(run, phase, gen_id):
+        f_works, r_works = _get_works(work, RUN, projects[phase], GEN=f"GEN{gen_id}")
+        f_works = _strip_outliers(f_works)
+        r_works = _strip_outliers(r_works)
+        fes = []
+        errs = []
+
+        if len(f_works) > 10 and len(r_works) > 10:
+            for _ in range(n_bootstrap):
+                f = random.choices(f_works, k=len(f_works))
+                r = random.choices(r_works, k=len(r_works))
+                fe, err = BAR(np.asarray(f), np.asarray(r))
+                fes.append(fe)
+                errs.append(err)
+
+        return fes, errs, f_works, r_works
+
+    if cache_dir is not None:
+        # cache results in local directory
+        memory = joblib.Memory(backend="local", location=cache_dir, verbose=0)
+        bootstrap_BAR = memory.cache(_bootstrap_BAR)
+    else:
+        bootstrap_BAR = _bootstrap_BAR
+
+    for d in tqdm.tqdm(details.values()):
```

Comment:
Use `from tqdm.auto import tqdm` so it auto-detects the frontend (ASCII on the command-line, widgets on the notebook).

Reply:
Cool! I didn't know about `tqdm.auto`

Reply:
4763bde